File: platform/bb/rhodes/platform/5.0/com/rho/db/SqliteStorage.java
/*------------------------------------------------------------------------
* (The MIT License)
* 
* Copyright (c) 2008-2011 Rhomobile, Inc.
* 
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
* 
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
* 
* http://rhomobile.com
*------------------------------------------------------------------------*/

package com.rho.db;

import j2me.lang.CharacterMe;

//import com.rho.AppBuildConfig;
//import com.rho.RhoConf;
import com.rho.RhoEmptyLogger;
import com.rho.RhoLogger;
import com.rho.file.*;

import net.rim.device.api.io.URI;
import net.rim.device.api.database.*;
import java.util.Vector;
//import java.util.Hashtable;
//import java.util.Enumeration;

public class SqliteStorage implements IDBStorage 
{
	private static final RhoLogger LOG = RhoLogger.RHO_STRIP_LOG ? new RhoEmptyLogger() : 
		new RhoLogger("SqliteStorage");
	
	private IFileAccess m_fs;
	private boolean m_bPendingTransaction = false;
	private Database m_db;
	private int m_nInsideTransaction = 0;
	private boolean m_bNeedProcessCallback = false;
	private IDBCallback m_dbCallback;
	//private Hashtable m_mapStatements = new Hashtable();
	
	public SqliteStorage()
	{
		m_fs = new FileAccessBB();
	}
	
	public void deleteAllFiles(String strPath) throws Exception 
	{
		String strDbName = getNameNoExt(strPath);
		m_fs.delete(strDbName + ".data");
		m_fs.delete(strDbName + ".data-journal");

		//hsql old files
		m_fs.delete(strDbName + ".script");
		m_fs.delete(strDbName + ".script.new");
		m_fs.delete(strDbName + ".journal");
		m_fs.delete(strDbName + ".properties");		
	}

	public IDBResult createResult() {
		return new SqliteResult();
	}

	void bindObject(Statement st, int i, Object obj )throws DatabaseException
	{
		if ( obj instanceof String)
			st.bind(i, ((String)obj));
		else if ( obj instanceof Integer)
			st.bind(i, ((Integer)obj).intValue());
		else if ( obj instanceof Short)
			st.bind(i, ((Short)obj).shortValue());
		else if ( obj instanceof Long)
			st.bind(i, ((Long)obj).longValue());
		else if ( obj instanceof Float)
			st.bind(i, ((Float)obj).floatValue());
		else if ( obj instanceof Double)
			st.bind(i, ((Double)obj).doubleValue());
		else if ( obj instanceof Byte)
			st.bind(i, ((Byte)obj).byteValue());
		else if ( obj instanceof byte[])
			st.bind(i, (byte[])obj );
		
	}

	IDBResult executeSQL(String strStatement, Object[] values) throws DBException 
	{
		return executeSQL(strStatement, values, false, false); 
	}
	
	public IDBResult executeSQL(String strStatement, Object[] values, boolean bReportNonUnique) throws DBException 
	{
		return executeSQL(strStatement, values, bReportNonUnique, false); 
	}
	
	public IDBResult executeSQL(String strStatement, Object[] values,
			boolean bReportNonUnique, boolean bNoCopy) throws DBException 
	{
		//LOG.INFO(strStatement);// + "; Values: " + values);
		synchronized(m_db)
		{
			if ( m_db == null )
				throw new RuntimeException("executeSQL: m_db == null");
	
			IDBResult res = null;
			
			String strStatementOrig = strStatement;
			
			try
			{
				while( strStatement != null && strStatement.length()> 0)
				{
		            int start = 0;
		            while ( start < strStatement.length() && (strStatement.charAt(start) == '\n' ||
		            		strStatement.charAt(start) == '\r' ||
		            		strStatement.charAt(start) == ';' ||
		            		CharacterMe.isWhitespace(strStatement.charAt(start) ) ) )
		            		start++;
		            if (start > 0 )
		            	strStatement = strStatement.substring(start);
		            	
					if ( strStatement == null || start >= strStatement.length() )
						break;
					
					String strCommand = strStatement.length() > 6 ? strStatement.substring(0, 6) : ""; 
					boolean bSelect = strCommand.equalsIgnoreCase("SELECT") || strCommand.equalsIgnoreCase("PRAGMA");
					
					Statement st = null;//(Statement)m_mapStatements.get(strStatement);
					boolean bCachedStatement = st != null;
					if ( st == null )
						st = m_db.createStatement(strStatement);
					
					boolean bDontCloseStatement = false;
					try
					{
						if ( !bCachedStatement )
						{
							st.prepare();
		                	strStatement = st.getTail();
		                	
		                	/*if ( strStatement == null || strStatement.length() == 0 )
		                	{
		                		m_mapStatements.put(strStatementOrig, st);
		                		bCachedStatement = true;
		                	}*/
		                	
						}else
							strStatement = "";
						
						for ( int i = 0; values != null && i < values.length; i++ )
						{
							bindObject(st, i+1, values[i]);
						}
						
						if ( bSelect )
						{
			                if ( res == null )
			                {
			                	res = new SqliteResult(st, bCachedStatement, bNoCopy);
			                	bDontCloseStatement = true;
			                }
						}else
						{
							try
							{
								//LOG.INFO("START execute");
								st.execute();
								//LOG.INFO("END execute");
								
								if ( /*strCommand.equalsIgnoreCase("INSERT")||*/ strCommand.equalsIgnoreCase("DELETE") ||
									 strCommand.equalsIgnoreCase("UPDATE")  )
								{
									if ( m_nInsideTransaction == 0 )
										processCallbackData();
									else
										m_bNeedProcessCallback = true;
								}
								
							}catch(DatabaseException exc)
							{
								if ( res == null && bReportNonUnique && exc.getMessage().indexOf("constraint failed") >= 0)
									res = new SqliteResult(true);
								else
									throw exc;
							}
							
		                	if ( res == null )
		                		res = new SqliteResult(null, false, false);
						}
						
					}finally
					{
						if ( !bDontCloseStatement )
						{
							if ( bCachedStatement )
							{
								//LOG.INFO("START reset");							
								st.reset();
								//LOG.INFO("END reset");
							}
							else
								st.close();
						}
						
						st = null;
					}
				}
			}catch(DatabaseException exc )
			{
				LOG.ERROR("executeSQL failed. Statement: " + strStatementOrig, exc);
				throw new DBException(exc);
			}
			
			//LOG.INFO("executeSQL END");
			return res;
		}
	}

	public boolean isTableExists(String strName)throws DBException
	{
		Object vals[] = {strName};
		IDBResult res = executeSQL("SELECT name FROM sqlite_master WHERE type='table' AND name=?", vals );
		boolean bRes = !res.isEnd();
		res.close();
		
		return bRes;
	}
	
	public String[] getAllTableNames() throws DBException 
	{
		IDBResult res = executeSQL("SELECT name FROM sqlite_master WHERE type='table'", null );
		
		Vector arTables = new Vector();
	    for ( ; !res.isEnd(); res.next() )
	    {
	    	arTables.addElement(res.getCurData()[0]);
	    }
	    
		String[] vecTables = new String[arTables.size()]; 
		for ( int i = 0; i<arTables.size();i++)
			vecTables[i] = (String)arTables.elementAt(i);
		
	    return vecTables;
	}

	public boolean isDbFileExists(String strPath) 
	{
		String strDbName = getNameNoExt(strPath);
		return m_fs.exists(strDbName + ".data");
	}

	private String getNameNoExt(String strPath){
		int nDot = strPath.lastIndexOf('.');
		String strDbName = "";
		if ( nDot > 0 )
			strDbName = strPath.substring(0, nDot);
		else
			strDbName = strPath;
		
		return strDbName;
	}
	
	public void executeBatchSQL(String strSqlScript)throws DBException
	{
		executeSQL(strSqlScript, null);
	}
	
	public static void OnInsertObjectRecord(Object NEW_source_id, Object NEW_attrib )
	{
		LOG.INFO("OnInsertObjectRecord");
	}
	
	public void createTriggers() throws DBException
	{
		String strTriggers = RhoFile.readStringFromJarFile("apps/db/syncdb_java.triggers", this); 
		executeBatchSQL( strTriggers );
	}
	
	public void open(String strPath, String strSqlScript, String strEncryptionInfo) throws DBException 
	{
		try{
			String strDbName = getNameNoExt(strPath) + ".data";
			
			//m_dbSess.setDBCallback(this);

			URI myURI = URI.create(strDbName);
			boolean bEncrypted =  strEncryptionInfo != null && strEncryptionInfo.length() >0;
			DatabaseSecurityOptions dbso = new DatabaseSecurityOptions(bEncrypted);
			
			if ( !m_fs.exists(strDbName) )
			{
				m_db = DatabaseFactory.create(myURI, dbso);
				
				m_db.beginTransaction();
				m_nInsideTransaction++;
				try
				{
					executeBatchSQL( strSqlScript );
					createTriggers();
				}catch(DBException exc)
				{
					m_db.rollbackTransaction();
					throw exc;
				}
				m_db.commitTransaction();
			}else
			{
				m_db = DatabaseFactory.openOrCreate(myURI, dbso);
			}
			
			if ( m_bPendingTransaction )
				m_db.beginTransaction();
			
			m_bPendingTransaction = false;
		}catch(Exception exc ){
			throw new DBException(exc);
		}finally
		{
			m_nInsideTransaction = 0;
		}
	}

	public void startTransaction() throws DBException 
	{
		synchronized(m_db)
		{
			try{
				//LOG.INFO("startTransaction START");
				if ( m_db == null )
					m_bPendingTransaction = true;
				else
					m_db.beginTransaction();
				
				m_nInsideTransaction++;
				m_bNeedProcessCallback = false;
				
				//LOG.INFO("startTransaction END");
			}catch(DatabaseException exc ){
				throw new DBException(exc);
			}
		}
	}
	
	public void onBeforeCommit() throws DBException
	{
		if ( !m_bNeedProcessCallback )
			return;
		
		//LOG.INFO("onBeforeCommit START");
		processCallbackData();
		//LOG.INFO("onBeforeCommit END");
	}
	
	public void commit() throws DBException 
	{
		synchronized(m_db)
		{
			try{
				//LOG.INFO("commit START");
				if ( m_db!= null )
					m_db.commitTransaction();
				//LOG.INFO("commit END");
			}catch(DatabaseException exc ){
				throw new DBException(exc);
			}finally
			{
				if ( m_nInsideTransaction > 0 )
					m_nInsideTransaction--;
			}
		}
	}
	
	public void rollback() throws DBException 
	{
		synchronized(m_db)
		{
			try{
				if ( m_db!= null )
					m_db.rollbackTransaction();
			}catch(DatabaseException exc ){
				throw new DBException(exc);
			}finally
			{
				if ( m_nInsideTransaction > 0 )
					m_nInsideTransaction--;
			}
		}
	}

	public void close() throws DBException 
	{
		synchronized(m_db)
		{
			try
			{
				/*Enumeration values = m_mapStatements.elements();
				while( values.hasMoreElements() )
				{
					Statement st = (Statement)values.nextElement();
					try{
						if ( st != null )
							st.close();
					}catch( DatabaseException exc  )
					{
						LOG.ERROR("close statement failed.", exc);
					}
				}*/
				if ( m_db!= null )
					m_db.close();
				
				m_db = null;
			}catch(DatabaseException exc ){
				throw new DBException(exc);
			}
		}
	}
	
	void processCallbackData() throws DBException
	{
		if ( m_dbCallback == null )
			return;
/*		{
			IDBResult rows2Insert = executeSQL("SELECT * FROM object_attribs_to_insert", null, false);
			if ( rows2Insert == null || rows2Insert.isEnd() )
			{
				if (rows2Insert != null)
					rows2Insert.close();
			}else
			{
				m_dbCallback.onAfterInsert("object_values", rows2Insert);
				
				m_nInsideTransaction++;
				try{
					executeSQL("DELETE FROM object_attribs_to_insert", null, false);
				}finally
				{
					m_nInsideTransaction--;
				}
			}
		}*/
		
		{
			IDBResult rows2Delete = executeSQL("SELECT * FROM object_attribs_to_delete", null);
			if ( rows2Delete == null || rows2Delete.isEnd() )
			{
				if (rows2Delete != null)
					rows2Delete.close();
			}else
			{
				m_dbCallback.onBeforeDelete("object_values", rows2Delete);
				
				m_nInsideTransaction++;
				try{
					executeSQL("DELETE FROM object_attribs_to_delete", null);
				}finally
				{
					m_nInsideTransaction--;
				}
			}
		}
		
		{
			IDBResult rows2Update = executeSQL("SELECT * FROM object_attribs_to_update", null);
			if ( rows2Update == null || rows2Update.isEnd() )
			{
				if (rows2Update != null)
					rows2Update.close();
			}else
			{
				int cols[] = {3};
				m_dbCallback.onBeforeUpdate("object_values", rows2Update, cols);
				
				m_nInsideTransaction++;
				try{
					executeSQL("DELETE FROM object_attribs_to_update", null);
				}finally
				{
					m_nInsideTransaction--;
				}
			}
		}
		
	}
	
	public void setDbCallback(IDBCallback callback) 
	{
		m_dbCallback = callback;
	}

}

File: platform/wp7/RhoRubyLib/db/SqliteStorage.cs
﻿/*------------------------------------------------------------------------
* (The MIT License)
* 
* Copyright (c) 2008-2011 Rhomobile, Inc.
* 
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
* 
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
* 
* http://rhomobile.com
*------------------------------------------------------------------------*/

using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using Community.CsharpSqlite;
using System.Globalization;
using IronRuby.Runtime;
using IronRuby.Builtins;
using rho.common;

namespace rho.db
{
    public class CSqliteStorage : IDBStorage
    {
        private static RhoLogger LOG = RhoLogger.RHO_STRIP_LOG ? new RhoEmptyLogger() :
            new RhoLogger("SqliteStorage");

        private boolean m_bPendingTransaction = false;
        private Sqlite3.sqlite3 m_db;
        private int m_nInsideTransaction = 0;
        private boolean m_bNeedProcessCallback = false;
        private IDBCallback m_dbCallback;

        public void close()
        {
            lock (m_db)
            {
                if (m_db != null)
                {
                    int rc = Sqlite3.sqlite3_close(m_db);
                    checkError(rc);
                    m_db = null;
                }
            }
        }

        public void commit()
        {
            lock (m_db)
            {
                if (m_nInsideTransaction > 0)
                    m_nInsideTransaction--;

                if (m_db != null)
                {
                    Sqlite3.sqlite3_exec(m_db, "END;", 0, 0, 0);
                    checkError();
                }
            }
        }

        public IDBResult createResult()
        {
            return new CSqliteResult();
        }

        public void createTriggers()
        {
            String strTriggers = CRhoFile.readStringFromResourceFile("db/syncdb.triggers");
            executeBatchSQL(strTriggers);
        }

        public void deleteAllFiles(String strPath)
        {
            CRhoFile.deleteFile(strPath);
            CRhoFile.deleteFile(strPath + "-journal");
        }

        public void executeBatchSQL(String strStatement)
        {
            lock (m_db)
            {
                Sqlite3.sqlite3_exec(m_db, strStatement, 0, 0, 0);
                checkError();
            }
        }

        void checkError()
        {
            int res = Sqlite3.sqlite3_errcode(m_db);
            checkError(res);
        }

        void checkError(int res)
        {
            if (res != Sqlite3.SQLITE_OK)
                throw new DBException(res, Sqlite3.sqlite3_errmsg(m_db));
        }

        public IDBResult executeSQL(String strStatement, Object[] values, boolean bReportNonUnique, boolean bNoCopy)
        {
            if (m_db == null)
                throw new Exception("executeSQL: m_db == null");

            lock (m_db)
            {
                var stmt = Prepare(strStatement, values);
                CSqliteResult res = new CSqliteResult(stmt, bNoCopy);
                int rc = res.executeStatement();
                if (rc != Sqlite3.SQLITE_ROW)
                {
                    if (bReportNonUnique && rc == Sqlite3.SQLITE_CONSTRAINT)
                    {
                        res.close();
                        return new CSqliteResult(true);
                    }

                    res.checkError();
                    res.close();
                }
                return res;
            }

        }

        public string[] getAllTableNames()
        {
            IDBResult res = executeSQL("SELECT name FROM sqlite_master WHERE type='table'", null, false, false);

            Vector<Object> arTables = new Vector<Object>();
            for (; !res.isEnd(); res.next())
            {
                arTables.addElement(res.getCurData()[0]);
            }

            String[] vecTables = new String[arTables.size()];
            for (int i = 0; i < arTables.size(); i++)
                vecTables[i] = (String)arTables.elementAt(i);

            return vecTables;
        }

        public boolean isDbFileExists(string strPath)
        {
            return CRhoFile.isFileExist(strPath);
        }
        
        public boolean isTableExists(string strName)
        {
            Object[] vals = { strName };
		    IDBResult res = executeSQL("SELECT name FROM sqlite_master WHERE type='table' AND name=?", vals, false, false );
		    boolean bRes = !res.isEnd();
		    res.close();
		
		    return bRes;
        }

        public void onBeforeCommit()
        {
            if (!m_bNeedProcessCallback)
                return;
            processCallbackData();
        }

        public void open(String strPath, String strSqlScript, String strEncryptionInfo)
        {
            try
            {
                String dbURI = RHODESAPP().resolveDBFilesPath(strPath);
                dbURI = CFilePath.removeFirstSlash(dbURI);
                boolean bEncrypted = strEncryptionInfo != null && strEncryptionInfo.length() > 0;
                //DatabaseSecurityOptions dbso = new DatabaseSecurityOptions(bEncrypted);

                boolean bExist = CRhoFile.isFileExist(strPath);

                int res = Sqlite3.sqlite3_open(dbURI, ref m_db);
                if (res != Sqlite3.SQLITE_OK)
                    throw new DBException(res, "Could not open database file: " + strPath);

                res = Sqlite3.sqlite3_exec(m_db, "PRAGMA journal_mode=PERSIST", 0, 0, 0);
                if (res != Sqlite3.SQLITE_OK)
                {
                    Sqlite3.sqlite3_close(m_db);
                    m_db = null;
                    throw new DBException(res, "Cannot set journal mode to PERSIST: " + strPath);
                }

                string[] ar1 = CRhoFile.enumDirectory("db");

                if (!bExist)
                    createSchema(strSqlScript);

                Sqlite3.sqlite3_create_function(m_db, "rhoOnDeleteObjectRecord", 3, Sqlite3.SQLITE_ANY, 0,
                                                DBAdapter.SyncBlob_DeleteCallback, null, null);
                Sqlite3.sqlite3_create_function(m_db, "rhoOnUpdateObjectRecord", 3, Sqlite3.SQLITE_ANY, 0,
                                                DBAdapter.SyncBlob_UpdateCallback, null, null);

                string[] ar2 = CRhoFile.enumDirectory("db");

                if (m_bPendingTransaction)
                    startTransaction();

                m_bPendingTransaction = false;
            }
            catch (Exception exc)
            {
                throw new DBException(exc);
            }
            finally
            {
                m_nInsideTransaction = 0;
            }
        }

        public void rollback()
        {
            if (m_nInsideTransaction > 0)
                m_nInsideTransaction--;

            lock (m_db)
            {
                if (m_db != null)
                {
                    Sqlite3.sqlite3_exec(m_db, "ROLLBACK;", 0, 0, 0);
                    checkError();
                }
            }
        }

        public void setDbCallback(IDBCallback callback)
        {
            m_dbCallback = callback;
        }

        public void startTransaction()
        {
            lock (m_db)
            {
                if (m_db == null)
                    m_bPendingTransaction = true;
                else
                {
                    Sqlite3.sqlite3_exec(m_db, "BEGIN IMMEDIATE;", 0, 0, 0);
                    checkError();
                }

                m_nInsideTransaction++;
                m_bNeedProcessCallback = false;
            }
        }

        #region Helpers

        private void createSchema(String strSqlScript)
        {
            if ( strSqlScript.length() == 0 )
            {
                LOG.ERROR("createSchema failed. Cannot read schema: " + strSqlScript);
                return;
            }

            executeBatchSQL(strSqlScript);
            createTriggers();
        }

        private void Bind(Sqlite3.Vdbe stmt, Object[] values)
        {
            for (int i = 0; values != null && i < values.Length; i++)
            {
                var obj = values[i];
                if (obj == null)
                    Sqlite3.sqlite3_bind_null(stmt, i + 1);
                else if (obj is Byte || obj is UInt16 || obj is SByte || obj is Int16 || obj is Int32)
                    Sqlite3.sqlite3_bind_int(stmt, i + 1, Convert.ToInt32(obj, CultureInfo.InvariantCulture));
                else if (obj is UInt32 || obj is Int64)
                    Sqlite3.sqlite3_bind_int64(stmt, i + 1, Convert.ToInt64(obj, CultureInfo.InvariantCulture));
                else if (obj is Single || obj is Double || obj is Decimal)
                    Sqlite3.sqlite3_bind_double(stmt, i + 1, Convert.ToDouble(obj, CultureInfo.InvariantCulture));
                else if (obj is String)
                    Sqlite3.sqlite3_bind_text(stmt, i + 1, (String)obj, ((String)obj).Length, null);
                else if (obj is MutableString)
                    Sqlite3.sqlite3_bind_text(stmt, i + 1, obj.ToString(), ((MutableString)obj).Length, null);
                else if (obj is byte[])
                    Sqlite3.sqlite3_bind_blob(stmt, i + 1, (byte[])obj, ((byte[])obj).Length, null);
                else if (obj is RubySymbol)
                {
                    String val = ((RubySymbol)obj).ToString();
                    Sqlite3.sqlite3_bind_text(stmt, i + 1, val, val.Length, null);
                }
                else if (obj is Boolean)
                {
                    String val = ((Boolean)obj) ? "true" : "false";
                    Sqlite3.sqlite3_bind_text(stmt, i + 1, val, val.Length, null);
                }
                else
                {
                    String val = obj.ToString();
                    Sqlite3.sqlite3_bind_text(stmt, i + 1, val, val.Length, null);
                }

                checkError();
            }
        }

        private Sqlite3.Vdbe Prepare(String strStatement, Object[] values)
        {
            Sqlite3.Vdbe ppStmt=new Sqlite3.Vdbe();
            Sqlite3.sqlite3_prepare_v2(m_db, strStatement, strStatement.Length, ref ppStmt, 0);
            checkError();
            
            Bind(ppStmt, values);
            return ppStmt;
        }

        private void processCallbackData()
    	{
        }

        private static CRhodesApp RHODESAPP() { return CRhodesApp.Instance; }

        #endregion
    }
}
